### PR TITLE
🐛 Requeue after 20 seconds if ControlPlane is not initialized yet

### PR DESCRIPTION
**Pull request by Unix4ever**
There was a fix https://github.com/kubernetes-sigs/cluster-api/pull/4947 a while ago, which has removed `Requeue: true`
from the return result in the machine controller.

While this change makes sense, it actually makes this reconcilation
loop pretty fragile: if the machines are not updated after `initialized`
is set in the control plane provier, the machine controller never starts
workload tracker and never updates the machine to be ready.

We have this issue when TalosControlPlane contoller runs along with
Sidero infrastructure provider.

Signed-off-by: Artem Chernyshev <artem.chernyshev@talos-systems.com>

### CODE REVIEW

**Comment by fabriziopandini** [comment]
> it actually makes this reconcilation loop pretty fragile: if the machines are not updated after initialized
> is set in the control plane provider, the machine controller never starts
> workload tracker and never updates the machine to be ready.

Trying to understand this point a little bit better
As far as I remember it is not required that every machine gets updated to setup a Tracker because:

- Machine Reconciler is watching for Cluster, and this ensures reconcile is triggered for each machine when control plane initialized condition bubbles up (when a Cluster is updated, all machines are reconciled); 
- As soon as the first machine gets processed, the Tracker for the newly available Cluster is started, and then it starts watching for all the nodes in the target Cluster (no need to start it on a machine basis; in fact, as soon as the Tracker exists, watchClusterNodes becomes a no-op).

Is there a use case not covered by ^^? Could you provide some more details?

File: controllers/machine_controller.go [hunk 1]
```diff
@@ -235,11 +235,14 @@ func patchMachine(ctx context.Context, patchHelper *patch.Helper, machine *clust
...
 	log := ctrl.LoggerFrom(ctx)
 
+	res := ctrl.Result{}
 	if conditions.IsTrue(cluster, clusterv1.ControlPlaneInitializedCondition) {
 		if err := r.watchClusterNodes(ctx, cluster); err != nil {
 			log.Error(err, "error watching nodes on target cluster")
 			return ctrl.Result{}, err
 		}
+	} else {
+		res.RequeueAfter = 20 * time.Second
```
Comment (vincepri) [inline controllers/machine_controller.go:245]: Why is this needed? The machine controller watches all Cluster objects and enqueue notifications for each Machine in a cluster if needed https://github.com/kubernetes-sigs/cluster-api/pull/5884/files#diff-871c17903b2fd6d91b1f8e5dd5704e6788c0f47131faaa5800ea946fd5ceff40R107

**Comment by enxebre** [comment]
>func ClusterToObjectsMapper(c client.Client, ro client.ObjectList, scheme *runtime.Scheme) (handler.MapFunc, error) {

As mentioned in the comments above the Machine controller watches the cluster CR and it should reconcile machines and set `watchClusterNodes` anytime `ControlPlaneInitializedCondition` transitions into true. If this is not happening this might be a red herring for an underlying root problem causing the issue. What do the logs say? Might your Cluster CR be filtered out by the machine controller for any reason?

**Comment by Unix4ever** [comment]
> Is there a use case not covered by ^^? Could you provide some more details?

I think you're right. Reconcile on cluster updates doesn't work in our case. I wonder if there something we don't set up properly in our providers.

> Might your Cluster CR be filtered out by the machine controller for any reason?

That very well can be. I'll try to dig into it more.

I'll close this PR.
